flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/Rowtime.scala

```diff
@@ -0,0 +1,131 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.descriptors
+
+import org.apache.flink.table.api.Types
+import org.apache.flink.table.descriptors.NormalizedProperties.{normalizeTimestampExtractor, normalizeWatermarkStrategy}
+import org.apache.flink.table.sources.tsextractors.{ExistingField, StreamRecordTimestamp, TimestampExtractor}
+import org.apache.flink.table.sources.wmstrategies.{AscendingTimestamps, BoundedOutOfOrderTimestamps, PreserveWatermarks, WatermarkStrategy}
+
+import scala.collection.mutable
+
+/**
+  * Rowtime descriptor for describing an event time attribute in the schema.
+  */
+class Rowtime extends Descriptor {
+
+  private var timestampExtractor: Option[TimestampExtractor] = None
+  private var watermarkStrategy: Option[WatermarkStrategy] = None
+
+  /**
+    * Sets a built-in timestamp extractor that converts an existing [[Long]] or
+    * [[Types.SQL_TIMESTAMP]] field into the rowtime attribute.
+    *
+    * @param fieldName The field to convert into a rowtime attribute.
+    */
+  def timestampFromField(fieldName: String): Rowtime = {
+    timestampExtractor = Some(new ExistingField(fieldName))
+    this
+  }
+
+  /**
+    * Sets a built-in timestamp extractor that converts the assigned timestamp from
+    * a DataStream API record into the rowtime attribute.
+    *
+    * Note: This extractor only works in streaming environments.
+    */
+  def timestampFromDataStream(): Rowtime = {
+    timestampExtractor = Some(new StreamRecordTimestamp)
+    this
+  }
+
+  /**
+    * Sets a custom timestamp extractor to be used for the rowtime attribute.
+    *
+    * @param extractor The [[TimestampExtractor]] to extract the rowtime attribute
+    *                  from the physical type.
+    */
+  def timestampFromExtractor(extractor: TimestampExtractor): Rowtime = {
+    timestampExtractor = Some(extractor)
+    this
+  }
+
+  /**
+    * Sets a built-in watermark strategy for ascending rowtime attributes.
+    *
+    * Emits a watermark of the maximum observed timestamp so far minus 1.
+    * Rows that have a timestamp equal to the max timestamp are not late.
+    */
+  def watermarkPeriodicAscending(): Rowtime = {
+    watermarkStrategy = Some(new AscendingTimestamps)
+    this
+  }
+
+  /**
+    * Sets a built-in watermark strategy for rowtime attributes which are out-of-order by a bounded
+    * time interval.
+    *
+    * Emits watermarks which are the maximum observed timestamp minus the specified delay.
+    */
+  def watermarkPeriodicBounding(delay: Long): Rowtime = {
+    watermarkStrategy = Some(new BoundedOutOfOrderTimestamps(delay))
+    this
+  }
+
+  /**
+    * Sets a built-in watermark strategy which indicates the watermarks should be preserved from the
+    * underlying DataStream API.
+    */
+  def watermarkFromDataStream(): Rowtime = {
```

Comment:
`preserveSourceWatermarks()`

`DataStream` is only an internal aspect that's not visible when using table sources.